File: Lession 3/text.Editor.jl
# Julia implementation of the TextEditor

mutable struct TextEditor
    text::String
    undoStack::Vector{String}
    redoStack::Vector{String}

    function TextEditor()
        text = ""
        undoStack = [text]  # Initialize with the empty state
        redoStack = String[]
        new(text, undoStack, redoStack)
    end
end

function typeText(editor::TextEditor, newText::String)
    editor.text *= newText
    push!(editor.undoStack, editor.text)
    empty!(editor.redoStack)  # Clear the redo stack whenever new text is typed
end

function undo(editor::TextEditor)
    if length(editor.undoStack) > 1 # Ensure at least one state remains in the stack
        push!(editor.redoStack, pop!(editor.undoStack))
        editor.text = editor.undoStack[end]
    else
        println("Nothing to undo")
    end
end

function redo(editor::TextEditor)
    if !isempty(editor.redoStack)
        editor.text = pop!(editor.redoStack)
        push!(editor.undoStack, editor.text)
    else
        println("Nothing to redo")
    end
end

function printText(editor::TextEditor)
    println(editor.text)
end

# Main function to demonstrate the TextEditor
function main()
    editor = TextEditor()

    typeText(editor, "Hello")
    typeText(editor, " World")
    printText(editor) # Output: Hello World

    undo(editor)
    printText(editor) # Output: Hello

    redo(editor)
    printText(editor) # Output: Hello World

    redo(editor)
    printText(editor) # Output: Hello World (No change, as redo stack is empty)

    undo(editor)
    printText(editor) # Output: Hello

    undo(editor)
    printText(editor) # Output: (Empty string)

    undo(editor) # Nothing to undo
end

main()

File: Lession 3/textEditor.cpp
#include <iostream>
#include <stack>
#include <string>

class TextEditor
{
private:
    std::string text;
    std::stack<std::string> undoStack;
    std::stack<std::string> redoStack;

public:
    TextEditor()
    {
        text = "";
        undoStack.push(text); // Initialize with the empty state
    }

    void typeText(const std::string &newText)
    {
        text += newText;
        undoStack.push(text);
        while (!redoStack.empty())
        {
            redoStack.pop(); // Clear the redo stack whenever new text is typed
        }
    }

    void undo()
    {
        if (undoStack.size() > 1)
        { // Ensure at least one state remains in the stack
            redoStack.push(undoStack.top());
            undoStack.pop();
            text = undoStack.top();
        }
        else
        {
            std::cout << "Nothing to undo" << std::endl;
        }
    }

    void redo()
    {
        if (!redoStack.empty())
        {
            text = redoStack.top();
            redoStack.pop();
            undoStack.push(text);
        }
        else
        {
            std::cout << "Nothing to redo" << std::endl;
        }
    }

    void printText() const
    {
        std::cout << text << std::endl;
    }
};

int main()
{
    TextEditor editor;

    editor.typeText("Hello");
    editor.typeText(" World");
    editor.printText(); // Output: Hello World

    editor.undo();
    editor.printText(); // Output: Hello

    editor.redo();
    editor.printText(); // Output: Hello World

    editor.redo();
    editor.printText(); // Output: Hello World (No change, as redo stack is empty)

    editor.undo();
    editor.printText(); // Output: Hello

    editor.undo();
    editor.printText(); // Output: (Empty string)

    editor.undo(); // Nothing to undo

    return 0;
}
